### PR TITLE
Add client-side cab token generator

### PR DESCRIPTION
**Pull request by nbayati**
See: [go/client-side-cab-design-doc](http://goto.google.com/client-side-cab-design-doc)
This PR merges the feature branch **client-side-cab** to **main**.

### CODE REVIEW

**Comment by nbayati** [comment]
The coverage and duplication conditions are failing on Sonar because of the generated code in `ClientSideAccessBoundaryProto.java`. We can ignore these two failed checks. 

Screenshot 2025-01-24 at 9 34 15 AM
Screenshot 2025-01-29 at 10 14 26 AM

File: oauth2_http/java/com/google/auth/oauth2/OAuth2Utils.java [hunk 1]
```diff
@@ -70,21 +70,27 @@
 import java.util.Set;
 
 /** Internal utilities for the com.google.auth.oauth2 namespace. */
-class OAuth2Utils {
+public class OAuth2Utils {
```
Comment (lqiu96) [inline oauth2_http/java/com/google/auth/oauth2/OAuth2Utils.java:73]: Since these were internal classes (package-private) and now public, thoughts on adding some strong-ish wording about how these are not intended for customers to use?

i.e. something
```
These classes are marked public but should treated effectively as internal classes only. These are not subject to any backwards compatibility requirements.
```

**Comment by lqiu96** [comment]
nit: Could we update the PR title to reflect the changes that are coming into main. We try use the PR titles to create release notes for each release.

> feat: add client-side cab token generator

Thoughts on something like `feat: Introduce Client-Side Certificate Access Boundary (CAB) functionality`

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java [hunk 2]
```diff
@@ -0,0 +1,674 @@
...
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.time.Duration;
+import java.util.Base64;
+import java.util.Date;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import javax.annotation.Nullable;
+
+public class ClientSideCredentialAccessBoundaryFactory {
```
Comment (lsirac) [inline cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:82]: Let's add javadocs with examples on how to use etc

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java [hunk 3]
```diff
@@ -0,0 +1,674 @@
...
+
+  /**
+   * Generates a Client-Side CAB token given the {@link CredentialAccessBoundary}.
+   *
+   * @param accessBoundary The credential access boundary that defines the restrictions for the
+   *     generated CAB token.
+   * @return The Client-Side CAB token in an {@link AccessToken} object
+   * @throws IOException If an I/O error occurs while refreshing the source credentials
+   * @throws CelValidationException If the availability condition is an invalid CEL expression
+   * @throws GeneralSecurityException If an error occurs during encryption
```
Comment (lsirac) [inline cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:130]:
```suggestion
  /**
   * Generates a downscoped access token given the {@link CredentialAccessBoundary}.
   *
   * @param accessBoundary The credential access boundary that defines the restrictions for the
   *     generated CAB token.
   * @return The downscoped access token in an {@link AccessToken} object
   * @throws IOException If an I/O error occurs while refreshing the source credentials
   * @throws CelValidationException If the availability condition is an invalid CEL expression
   * @throws GeneralSecurityException If an error occurs during encryption
```

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java [hunk 4]
```diff
@@ -0,0 +1,674 @@
...
+
+    // Handle the refresh based on the determined refresh type.
+    switch (refreshType) {
+      case BLOCKING:
+        if (refreshTask.isNew) {
+          // Start a new refresh task only if the task is new
+          MoreExecutors.directExecutor().execute(refreshTask.task);
+        }
+        try {
+          refreshTask.task.get(); // Wait for the refresh task to complete.
```
Comment (lsirac) [inline cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:185]: nit: Let's be consistent with comments (also change elsewhere)

```suggestion
          // Wait for the refresh task to complete.
          refreshTask.task.get();
```

File: cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java [hunk 5]
```diff
@@ -0,0 +1,674 @@
...
+    }
+
+    // If a refresh is required, create or retrieve the refresh task.
+    RefreshTask refreshTask = getOrCreateRefreshTask();
+
+    // Handle the refresh based on the determined refresh type.
+    switch (refreshType) {
+      case BLOCKING:
+        if (refreshTask.isNew) {
+          // Start a new refresh task only if the task is new
```
Comment (lsirac) [inline cab-token-generator/java/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactory.java:181]:
```suggestion
          // Start a new refresh task only if the task is new.
```

File: cab-token-generator/javatests/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactoryTest.java [hunk 6]
```diff
@@ -0,0 +1,954 @@
+/*
+ * Copyright 2024, Google LLC
```
Comment (lsirac) [inline cab-token-generator/javatests/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactoryTest.java:2]:
```suggestion
 * Copyright 2025, Google LLC
```

File: cab-token-generator/javatests/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactoryTest.java [hunk 7]
```diff
@@ -0,0 +1,954 @@
...
+            NullPointerException.class,
+            () ->
+                ClientSideCredentialAccessBoundaryFactory.newBuilder()
+                    .setHttpTransportFactory(OAuth2Utils.HTTP_TRANSPORT_FACTORY)
+                    .build());
+    assertEquals("Source credential must not be null.", exception.getMessage());
+  }
+
+  @Test
+  public void builder_minimumTokenLifetime_negative_throws() throws IOException {
```
Comment (lsirac) [inline cab-token-generator/javatests/com/google/auth/credentialaccessboundary/ClientSideCredentialAccessBoundaryFactoryTest.java:394]: Add tests for asserting minimum token lifetime behavior & refresh margin behavior (I might've missed it so resolve if so)

File: cab-token-generator/javatests/com/google/auth/credentialaccessboundary/ITClientSideCredentialAccessBoundaryTest.java [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.auth.oauth2.OAuth2CredentialsWithRefresh;
+import com.google.auth.oauth2.ServiceAccountCredentials;
+import dev.cel.common.CelValidationException;
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import org.junit.Test;
+
+/**
+ * Integration tests for the Client Side Credential Access Boundary Factory.
```
Comment (lsirac) [inline cab-token-generator/javatests/com/google/auth/credentialaccessboundary/ITClientSideCredentialAccessBoundaryTest.java:58]: nit: just use @link here

File: pom.xml [hunk 9]
```diff
@@ -308,7 +354,8 @@
         <groupId>org.codehaus.mojo</groupId>
         <artifactId>clirr-maven-plugin</artifactId>
         <configuration>
-          <ignoredDifferencesFile>clirr-ignored-differences.xml</ignoredDifferencesFile>
```
Comment (lqiu96) [inline pom.xml:311]: Is this still needed?

**Review by lqiu96 — APPROVED** [review]

**Review by lsirac — APPROVED** [review]

**Review by aeitzman — APPROVED** [review]
